### PR TITLE
refactor(number): convert files to typescript (FE-4813)

### PR DESCRIPTION
**Pull request by grabkowski**
### Proposed behaviour

All component files should be refactored and converted to TypeScript.

### Current behaviour

Component files are currently javascript.

### Checklist

- [x] Commits follow our style guide
~~- [ ] Related issues linked in commit messages if required~~
~~- [ ] Screenshots are included in the PR if useful~~
~~- [ ] All themes are supported if required~~
- [x] Unit tests added or updated if required
~~- [ ] Cypress automation tests added or updated if required~~
- [x] Storybook added or updated if required
~~- [ ] Translations added or updated (including creating or amending translation keys table in storybook) if required~~
- [x] Typescript `d.ts` file added or updated if required

#### QA

- [ ] Tested in CodeSandbox/storybook
- [ ] Add new Cypress test coverage if required
- [ ] Carbon implementation matches Design System/designs
- [ ] UI Tests GitHub check reviewed if required

### Testing instructions

The following CodeSandbox is an example of the broken behaviour.
You can see the new behaviour by looking at the version in the comment by `codesandbox[bot]`.

### CODE REVIEW

File: src/components/number/number.spec.js [hunk 1]
```diff
@@ -6,8 +6,22 @@ import Label from "../../__internal__/label";
 import { assertStyleMatch } from "../../__spec_helper__/test-utils";
 import InputPresentation from "../../__internal__/input/input-presentation.component";
 
+function renderNumberInput(props, renderer = mount) {
```
Comment (edleeks87) [inline src/components/number/number.spec.js]: We don't need the second arg here as mount is the only renderer used

File: src/components/number/number.component.tsx [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    <Textbox
+      {...rest}
+      value={value}
+      onChange={handleOnChange}
+      onKeyDown={handleKeyDown}
+    />
+  );
+};
+
+Number.displayName = "Number";
```
Comment (robinzigmond) [inline src/components/number/number.component.tsx:55]: do we really have to explicitly set the displayName here? I thought React would set that automatically, unless we were using forwardRef or something (which we're not, at least for now).

It won't do any harm (and may save us some work/bugs in future), I'm just surprised it's been added as part of a TS refactor.

File: src/components/number/number.spec.js (hunk 1 above)
Comment (grabkowski) [reply src/components/number/number.spec.js]: right, removing it now

File: src/components/number/number.component.tsx (hunk 2 above)
Comment (grabkowski) [reply src/components/number/number.component.tsx:55]: I'm pretty sure that the `Number` component should be wrapped in a `forwardRef`. I'll remove it.

**Review by robinzigmond — APPROVED** [review]

**Review by nineteen88 — DISMISSED** [review]

**Review by divyajindel — APPROVED** [review]

**Review by robinzigmond — APPROVED** [review]

**Review by nineteen88 — APPROVED** [review]

**Review by nineteen88 — APPROVED** [review]

**Review by robinzigmond — APPROVED** [review]

**Review by divyajindel — APPROVED** [review]

**Pull request merged**